### PR TITLE
Use early_inc_range

### PR DESCRIPTION
**Pull request by tgymnich**
Get rid of quite a few allocations by using `make_early_inc_range`

### CODE REVIEW

File: enzyme/Enzyme/TypeAnalysis/TypeTree.h [hunk 1]
```diff
@@ -248,7 +247,8 @@ class TypeTree : public std::enable_shared_from_this<TypeTree> {
             if (CT == BaseType::Anything || CT == pair.second) {
               // previous equivalent values or values overwritten by
               // an anything are removed
-              toremove.insert(pair.first);
+              changed = true;
```
Comment (wsmoses) [inline enzyme/Enzyme/TypeAnalysis/TypeTree.h:250]: Is this legal here (std map iirc), for some std containers I remember erasure invalidates all iterators.

**Comment by wsmoses** [comment]
This is fantastic and should be done throughout the codebase. However, I don't remember if the datastructre here (sstd::map) would support this.

File: enzyme/Enzyme/TypeAnalysis/TypeTree.h (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/TypeAnalysis/TypeTree.h:250]: > References and iterators to the erased elements are invalidated. Other references and iterators are not affected.
https://en.cppreference.com/w/cpp/container/map/erase

**Comment by wsmoses** [comment]
References and iterators to the erased elements are invalidated. Other references and iterators are not affected.

from `https://en.cppreference.com/w/cpp/container/map/erase`

So looks like this should be fine

**Review by wsmoses — APPROVED** [review]

**Comment by tgymnich** [comment]
> This is fantastic and should be done throughout the codebase. However, I don't remember if the datastructre here (sstd::map) would support this.
@wsmoses I've got some more of this lined up. Just testing the waters with this PR.

**Pull request merged**